### DISCUSSION TITLE
token in hostRules is not applied

### DISCUSSION THREAD

**Discussion by tuft89** [renovatebot/renovate#22037] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.73.0

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

3.8.2

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I'm using two github app installation tokens.
token "X" used for running renovate in orgA/projectA actions, and another token "Y" is used for accessing private package in orgB/private-package.
token "Y" is configured in hostRules like below, but renovate seem to access orgB/private-package with token "X".
```
"regexManagers": [
    {
      "fileMatch": ["src/.*py$"],
      "matchStrings": ["::(?<registryUrl>.*?)\\/(?<depname>.*?)\\?ref=(?<currentValue>.*?)\""],
      "packageNameTemplate": "orgB/private-package",
      "depNameTemplate": "orgB/private-package",
      "datasourceTemplate": "github-tags",
      "registryUrlTemplate": "https://my-github-enterprise-server.com"
    }
  ],
  "hostRules":[
    {
      "matchHost": "https://my-github-enterprise-server.com/orgB/private-package",
      "token": "process.env.RENOVATE_TOKEN_Y"
    }
```

### What you tried so far.

I have verified both tokens are valid with curl, and debug with vs code.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Found hostRules token for url https://my-github-enterprise-server.com/orgB/private-package (repository=orgA/projectA)
DEBUG: Datasource unknown error (repository=orgA/projectA)
       "datasource": "github-tags",
       "packageName": "orgB/private-package",
       "err": {
         "message": "Could not resolve to a Repository with the name 'orgB/private-package'.",
         "stack": "Error: Could not resolve to a Repository with the name 'orgB/private-package'.\n    at GithubGraphqlDatasourceFetcher.doRawQuery (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:132:21)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at GithubGraphqlDatasourceFetcher.doShrinkableQuery (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:197:20)\n    at GithubGraphqlDatasourceFetcher.doPaginatedQuery (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:237:27)\n    at GithubGraphqlDatasourceFetcher.getItems (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:286:17)\n    at Function.query (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:57:19)\n    at queryTags (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/index.ts:15:15)\n    at GithubTagsDatasource.getReleases (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/modules/datasource/github-tags/index.ts:68:24)\n    at getRegistryReleases (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/modules/datasource/index.ts:73:15)\n    at fetchReleases (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/modules/datasource/index.ts:299:15)"
       }
```

</details>

**Comment by rarkins** [comment]
`process.env.ANYTHING` should never be inside quotes, as you have above, otherwise it won't be interpolated by the environment variable. If the above is a `renovate.json` then it's impossible to interpolate

**Comment by tuft89** [comment]
Thank you @rarkins ,
I have replaced renovate.json with config.js and updated the configuration as follows.
```
  hostRules:[
    {
      matchHost: 'https://my-github-enterprise-server.com/orgB/private-package',
      token: process.env.RENOVATE_TOKEN_Y
    }
  ],
```

however, result is not changed. Is there anything else suspicious?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Found hostRules token for url https://my-github-enterprise-server.com/orgB/private-package (repository=orgA/projectA)
DEBUG: Datasource unknown error (repository=orgA/projectA)
       "datasource": "github-tags",
       "packageName": "orgB/private-package",
       "err": {
         "message": "Could not resolve to a Repository with the name 'orgB/private-package'.",
         "stack": "Error: Could not resolve to a Repository with the name 'orgB/private-package'.\n    at GithubGraphqlDatasourceFetcher.doRawQuery (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:132:21)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at GithubGraphqlDatasourceFetcher.doShrinkableQuery (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:197:20)\n    at GithubGraphqlDatasourceFetcher.doPaginatedQuery (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:237:27)\n    at GithubGraphqlDatasourceFetcher.getItems (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:286:17)\n    at Function.query (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/datasource-fetcher.ts:57:19)\n    at queryTags (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/util/github/graphql/index.ts:15:15)\n    at GithubTagsDatasource.getReleases (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/modules/datasource/github-tags/index.ts:68:24)\n    at getRegistryReleases (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/modules/datasource/index.ts:73:15)\n    at fetchReleases (/opt/buildpack/tools/renovate/35.73.0/node_modules/renovate/lib/modules/datasource/index.ts:299:15)"
       }
```

</details>

**Reply by rarkins** [reply]
Try setting hostType=github or hostType=github-tags in that host rule

**Reply by tuft89** [reply]
Thank you for thinking of me so many times @rarkins.
I have tried with both settings, but result is not changed.

**Reply by rarkins** [reply]
It probably needs something like this:

```
  hostRules:[
    {
      matchHost: 'https://my-github-enterprise-server.com/api/v3/repos/orgB/private-package',
      hostType: 'github',
      token: process.env.RENOVATE_TOKEN_Y
    }
  ],
```

i.e. the `/api/v3/repos` part added so that it matches the API.

**Reply by tuft89** [reply]
@rarkins, Thank you for your suggestion.
I have tried with `/api/v3/repos`, and error message was same but following debug log is not appeared.
```
DEBUG: Found hostRules token for url https://my-github-enterprise-server.com/orgB/private-package (repository=orgA/projectA)
```
I guess updated url was not found. 
Suggested path is rest , but I thought graphql is used for fetching packages.

**Reply by rarkins** [reply]
Graphql is used, but renovate searches the rest url for a matching host rule. Unfortunately it's very here for me to troubleshoot remotely like this

**Reply by tuft89** [reply]
Sorry for bother you, and really appreciate your help.
What information is needed for in-depth troubleshooting?

**Reply by rarkins** [reply]
Local debugging of Renovate from source. i.e. needs to be you or a colleague running `yarn debug` from source so you can step through the hostRules logic when the package is looked up

**Reply by tuft89** [reply]
Thank you @rarkins.
I have verified with local debugging, and find `token` and `matchHost` are overwrited by [isHostOnlyrule](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L82)  at [here](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L140).

token `Y` was replaced to token `X`, and matchHost was also replaced `https://my-github-enterprise-server.com/orgB/private-package` to `https://my-github-enterprise-server.com`.

**Reply by rarkins** [reply]
Are you setting hostType as I recommended earlier?

**Reply by tuft89** [reply]
sorry hostType was removed, so tried again with hostType.
`token` and `matchHost` was replaced by[ isMultiRule](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L86) at [here](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L149)

**Reply by rarkins** [reply]
What do you mean "replaced"? Do you mean filtered out? And if so, can you you work out whether it was the host or it was the hostType filter?

**Reply by tuft89** [reply]
Yes, I wanted to say filtered out. Let me explain precisely.
I'm checking variable with vs code debug console, and verified combination of `token` "Y" and `matchHost`  "https://my-github-enterprise-server.com/orgB/private-package" is exist at  [here](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L146).
`isMultiRule` return value was `true`,
`matchesHostType` return value was `false` , cause  `rule.hostType` is not match `search.hostType`. `rule.hostType` was `github`, and `seach.hostType` was `merge-confidence`.

**Reply by tuft89** [reply]
In the case `search.hostType` was `github`,  matchesHostType return value was `true`, however `matchesHost` failed at `validateUrl`.
`matchHost` value is 'https://my-github-enterprise-server.com/api/v3/repos/orgB/private-package', as you suggested.
I have verified with curl this url return 200 ok, but got return value `false` at [here](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L100).

**Reply by tuft89** [reply]
At this point, `seach` value was  not `https://my-github-enterprise-server.com/api/v3/repos` but  `https://my-github-enterprise-server.com/api/v3`.

**Reply by tuft89** [reply]
so I guess, root cause is missing `/repos` url.

**Reply by rarkins** [reply]
I'm not sure I follow. If the `search` value is `https://my-github-enterprise-server.com/api/v3` then it's not the right/relevant URL for your problem. Look for a search string which includes your org/repo

**Reply by tuft89** [reply]
Yes, My point was incorrect, `/repos` is not related the issue, sorry for my mistake.
Let me confirm another process at [here](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L100).
 In my environment, `search.url` were `https://my-github-enterprise-server.com/api/v3/` , `https://my-github-enterprise-server.com/api/graphql` , or `https://my-github-enterprise-server.com/api/v3/repos/repoA/projectA/pulls?per_page=100&state=all&sort=updated&direction=desc&page=1`.
`rule.matchHost` was `https://my-github-enterprise-server.com/orgB/private-package`.
I'm guessing  problem is  url `https://my-github-enterprise-server.com/orgB/private-package`  was not input as `search.url`.
